### PR TITLE
Fix symlink regression

### PR DESCRIPTION
**Pull request by robwise**
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/541)

### CODE REVIEW

**Comment by robwise** [comment]
Possibly related: https://github.com/shakacode/react_on_rails/issues/537#issuecomment-244143141

**Comment by dzirtusss** [comment]
Reviewed 3 of 3 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/react_on_rails/assets_precompile.rb, line 40 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQtmYRg5DNu6kPPCUKy:-KQtmYRg5DNu6kPPCUKz:bnrfy9x) ([raw file](https://github.com/shakacode/react_on_rails/blob/a17f7fd32d8ab7513c4930802c01f789f13691f4/lib/react_on_rails/assets_precompile.rb#L40)):*
> ```Ruby
>       if file_or_symlink_exists_at_path?(symlink_path)
>         puts "React On Rails: Removing existing invalid symlink or file #{symlink_path}"
>         `rm -f "#{symlink_path}"`
> ```

I think we should not use any OS calls in gem at all. Why not to change to `FileUtils` or smth? Some previous errors were because somebody tried to use filenames with spaces and all that process get broken.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/541)*

**Comment by justin808** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/react_on_rails/assets_precompile.rb, line 40 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQtmYRg5DNu6kPPCUKy:-KQttPX3lz7rDEmrna9d:b-y9nykk) ([raw file](https://github.com/shakacode/react_on_rails/blob/a17f7fd32d8ab7513c4930802c01f789f13691f4/lib/react_on_rails/assets_precompile.rb#L40)):*
<details><summary><i>Previously, dzirtusss (Sergey Tarasov) wrote…</i></summary>
> I think we should not use any OS calls in gem at all. Why not to change to `FileUtils` or smth? Some previous errors were because somebody tried to use filenames with spaces and all that process get broken.
</details>
I'm OK with that change, @dzirtusss 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/541)*

**Comment by robwise** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/react_on_rails/assets_precompile.rb, line 40 \[r2\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQtmYRg5DNu6kPPCUKy:-KQvFOUrenT5g9aPCXms:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/a17f7fd32d8ab7513c4930802c01f789f13691f4/lib/react_on_rails/assets_precompile.rb#L40)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary>
> I'm OK with that change, @dzirtusss 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/541)*

**Comment by justin808** [comment]
Overall good. Couple questions. Plus, can we get a couple unit tests that show the changes?

---

Reviewed 2 of 3 files at r1, 1 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[lib/react_on_rails/assets_precompile.rb, line 31 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQxKUcrJGl6jnnJEwlp:-KQxKUcrJGl6jnnJEwlq:b528x7d) ([raw file](https://github.com/shakacode/react_on_rails/blob/7ecdc36cb68d58d5ab9817455949fad962b0ebcb/lib/react_on_rails/assets_precompile.rb#L31)):*
> ```Ruby
>       # File.exist?(symlink_path) will check the file the sym is pointing to is existing
>       # File.lstat(symlink_path).symlink? confirms that this is a symlink
>       valid_symlink_already_exists = File.exist?(symlink_path) && File.lstat(symlink_path).symlink?
> ```

Can this be a method and end with `?`? and have a unit test?

---

*[lib/react_on_rails/assets_precompile.rb, line 85 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQxKt688WzlQvW9SBpy:-KQxKt688WzlQvW9SBpz:bl90hsq) ([raw file](https://github.com/shakacode/react_on_rails/blob/7ecdc36cb68d58d5ab9817455949fad962b0ebcb/lib/react_on_rails/assets_precompile.rb#L85)):*
> ```Ruby
> 
>           # We want the gz ones as well if they exist
>           if File.exist?(@assets_path.join("#{rails_digested_filename}.gz"))
> ```

Why do we need to check this here?

---

*[lib/react_on_rails/assets_precompile.rb, line 123 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQxKhbEwGv6_IyxBYjj:-KQxKhbEwGv6_IyxBYjk:b-bdyay2) ([raw file](https://github.com/shakacode/react_on_rails/blob/7ecdc36cb68d58d5ab9817455949fad962b0ebcb/lib/react_on_rails/assets_precompile.rb#L123)):*
> ```Ruby
>     private
> 
>     def file_or_symlink_exists_at_path?(path)
> ```

unit test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-:-KQxKy7o4U-8L-izMRZM:b-rjvk55)*

**Comment by robwise** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[lib/react_on_rails/assets_precompile.rb, line 31 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQxKUcrJGl6jnnJEwlp:-KR03eTYIQ8cVObrY1aL:b-onlhjt) ([raw file](https://github.com/shakacode/react_on_rails/blob/7ecdc36cb68d58d5ab9817455949fad962b0ebcb/lib/react_on_rails/assets_precompile.rb#L31)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary>
> Can this be a method and end with `?`? and have a unit test?
</details>
I didn't write this, I just renamed the variable

---

*[lib/react_on_rails/assets_precompile.rb, line 85 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQxKt688WzlQvW9SBpy:-KR03ADa6t2okSNFKfP8:b-ieudpa) ([raw file](https://github.com/shakacode/react_on_rails/blob/7ecdc36cb68d58d5ab9817455949fad962b0ebcb/lib/react_on_rails/assets_precompile.rb#L85)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary>
> Why do we need to check this here?
</details>
it'll throw if you try to symlink a file that doesn't exist, so all of the tests you wrote would fail without this

---

*[lib/react_on_rails/assets_precompile.rb, line 123 \[r3\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQxKhbEwGv6_IyxBYjj:-KR03PCAXvbQUlekskT-:b-vhfdmp) ([raw file](https://github.com/shakacode/react_on_rails/blob/7ecdc36cb68d58d5ab9817455949fad962b0ebcb/lib/react_on_rails/assets_precompile.rb#L123)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary>
> unit test?
</details>
you can't unit test private methods, not to mention I don't see the need to unit test whether core Ruby methods work

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/541)*

**Comment by justin808** [comment]
@robwise or @dzirtusss Please confirm my last commit. I changed the things I commented on.

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 1]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
           end.to raise_exception(AssetsPrecompile::SymlinkTargetDoesNotExistException)
         end
       end
+
+      it "creates a proper symlink when a file exists at destination" do
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:66]: you should specify which destination you are talking about

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 2]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
...
+        existing_filename = File.basename(Tempfile.new("tempfile", assets_path))
+        digest_filename = existing_filename
+
+        AssetsPrecompile.new(assets_path: assets_path).symlink_file(filename, digest_filename)
+        expect(assets_path.join(digest_filename).lstat.symlink?).to be true
+        expect(File.identical?(assets_path.join(filename),
+                               assets_path.join(digest_filename))).to be true
+      end
+
+      it "creates a proper symlink when a symlink file exists at destination" do
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:77]: is it a valid symlink or invalid symlink? It doesn't do anything if it's valid, just outputs a message

``` ruby
context "when there is already a valid symlink in place" do
   it "outputs a message saying that it need not perform any action" do
```

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 3]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
...
+        File.symlink(existing_filename, digest_filename)
+
+        AssetsPrecompile.new(assets_path: assets_path).symlink_file(filename, digest_filename)
+        expect(assets_path.join(digest_filename).lstat.symlink?).to be true
+        expect(File.identical?(assets_path.join(filename),
+                               assets_path.join(digest_filename))).to be true
+        File.delete(digest_filename)
+      end
+
+      it "creates a proper symlink when an invalid symlink exists at destination" do
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:93]:
``` ruby
context "when there is an existing invalid symlink" do
  it "replaces it with a valid one" do
```

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 4]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
...
+        digest_file = Tempfile.new("tempfile", assets_path)
+        digest_filename = File.basename(digest_file)
+        File.symlink(existing_filename, digest_filename)
+        File.delete(existing_file) # now digest_filename is an invalid link
+
+        AssetsPrecompile.new(assets_path: assets_path).symlink_file(filename, digest_filename)
+        expect(assets_path.join(digest_filename).lstat.symlink?).to be true
+        expect(File.identical?(assets_path.join(filename),
+                               assets_path.join(digest_filename))).to be true
+        File.delete(digest_filename)
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:107]: why are you doing this, it's a `TempFile`?

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 5]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
...
+        digest_file = Tempfile.new("tempfile", assets_path)
+        digest_filename = File.basename(digest_file)
+        File.delete(digest_file)
+        File.symlink(existing_filename, digest_filename)
+
+        AssetsPrecompile.new(assets_path: assets_path).symlink_file(filename, digest_filename)
+        expect(assets_path.join(digest_filename).lstat.symlink?).to be true
+        expect(File.identical?(assets_path.join(filename),
+                               assets_path.join(digest_filename))).to be true
+        File.delete(digest_filename)
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:90]: Unnecessary, this is a `TempFile`

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 6]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
...
+      it "creates a proper symlink when a symlink file exists at destination" do
+        filename = File.basename(Tempfile.new("tempfile", assets_path))
+        existing_filename = File.basename(Tempfile.new("tempfile", assets_path))
+        digest_file = Tempfile.new("tempfile", assets_path)
+        digest_filename = File.basename(digest_file)
+        File.delete(digest_file)
+        File.symlink(existing_filename, digest_filename)
+
+        AssetsPrecompile.new(assets_path: assets_path).symlink_file(filename, digest_filename)
+        expect(assets_path.join(digest_filename).lstat.symlink?).to be true
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:86]: usually we have a line between the action taken and the expectation

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 7]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
           end.to raise_exception(AssetsPrecompile::SymlinkTargetDoesNotExistException)
         end
       end
+
+      it "creates a proper symlink when a file exists at destination" do
+        filename = File.basename(Tempfile.new("tempfile", assets_path))
+        existing_filename = File.basename(Tempfile.new("tempfile", assets_path))
+        digest_filename = existing_filename
+
+        AssetsPrecompile.new(assets_path: assets_path).symlink_file(filename, digest_filename)
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:70]: usually we have a line between the action taken and the expectation

File: spec/react_on_rails/assets_precompile_spec.rb [hunk 8]
```diff
@@ -62,6 +62,48 @@ module ReactOnRails
...
+                               assets_path.join(digest_filename))).to be true
+      end
+
+      it "creates a proper symlink when a symlink file exists at destination" do
+        filename = File.basename(Tempfile.new("tempfile", assets_path))
+        existing_filename = File.basename(Tempfile.new("tempfile", assets_path))
+        digest_file = Tempfile.new("tempfile", assets_path)
+        digest_filename = File.basename(digest_file)
+        File.delete(digest_file)
+        File.symlink(existing_filename, digest_filename)
```
Comment (robwise) [inline spec/react_on_rails/assets_precompile_spec.rb:83]: I don't get what's going on here compared to your test description

**Comment by justin808** [comment]
Review status: 1 of 3 files reviewed at latest revision, 12 unresolved discussions.

---

*[spec/react_on_rails/assets_precompile_spec.rb, line 66 \[r4\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQsHg2j4_be1wmVaOqw-r4-66:-KR7Re9kyJ0ud0Re-jPw:b-5pzo0g) ([raw file](https://github.com/shakacode/react_on_rails/blob/168f3c8ba61c9a9b8d1a64e4e6f650bdef075b1d/spec/react_on_rails/assets_precompile_spec.rb#L66)):*
<details><summary><i>Previously, robwise (Rob Wise) wrote…</i></summary>
> you should specify which destination you are talking about
</details>
destination is where you want to create the symlink

what's a better word for this?

Here's the docs: 

```
symlink(old_name, new_name) → 0 click to toggle source
Creates a symbolic link called new_name for the existing file old_name. Raises a NotImplemented exception on platforms that do not support symbolic links.
```

suggestion?

---

*[spec/react_on_rails/assets_precompile_spec.rb, line 77 \[r4\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQsHg2j4_be1wmVaOqw-r4-77:-KR7S3vB_QlAQebiSphq:b-mo684g) ([raw file](https://github.com/shakacode/react_on_rails/blob/168f3c8ba61c9a9b8d1a64e4e6f650bdef075b1d/spec/react_on_rails/assets_precompile_spec.rb#L77)):*
<details><summary><i>Previously, robwise (Rob Wise) wrote…</i></summary>
> is it a valid symlink or invalid symlink? It doesn't do anything if it's valid, just outputs a message
> 
> 
> 
> ```ruby
> 
> context "when there is already a valid symlink in place" do
> 
>    it "outputs a message saying that it need not perform any action" do
> 
> ```
</details>
@robwise you're thinking of the calling code that does this check. This method doesn't care.

---

*[spec/react_on_rails/assets_precompile_spec.rb, line 83 \[r4\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQsHg2j4_be1wmVaOqw-r4-83:-KR7S9dKUQv8lsa2nVRL:b32vmt) ([raw file](https://github.com/shakacode/react_on_rails/blob/168f3c8ba61c9a9b8d1a64e4e6f650bdef075b1d/spec/react_on_rails/assets_precompile_spec.rb#L83)):*
<details><summary><i>Previously, robwise (Rob Wise) wrote…</i></summary>
> I don't get what's going on here compared to your test description
</details>
??

---

*[spec/react_on_rails/assets_precompile_spec.rb, line 89 \[r4\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQsHg2j4_be1wmVaOqw-r4-89:-KR7SCDJdaz3i7uavP3l:bygn4la) ([raw file](https://github.com/shakacode/react_on_rails/blob/168f3c8ba61c9a9b8d1a64e4e6f650bdef075b1d/spec/react_on_rails/assets_precompile_spec.rb#L89)):*
<details><summary><i>Previously, robwise (Rob Wise) wrote…</i></summary>
> Unnecessary, this is a `TempFile`
</details>
Not true in this case, as I created a symlink.

---

*[spec/react_on_rails/assets_precompile_spec.rb, line 105 \[r4\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/541#-KQsHg2j4_be1wmVaOqw-r4-105:-KR7SII3e62dNznDGCd7:bygn4la) ([raw file](https://github.com/shakacode/react_on_rails/blob/168f3c8ba61c9a9b8d1a64e4e6f650bdef075b1d/spec/react_on_rails/assets_precompile_spec.rb#L105)):*
<details><summary><i>Previously, robwise (Rob Wise) wrote…</i></summary>
> why are you doing this, it's a `TempFile`?
</details>
Not true in this case, as I created a symlink.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/541)*

**Pull request merged**